seccomp/xattr.go

```diff
@@ -397,23 +412,25 @@ func (si *listxattrSyscallInfo) processListxattr() (*sysResponse, error) {
 	resp := responseMsg.Payload.(domain.ListxattrRespPayload)
 
 	// Write the data returned by listxattr() to the memory of the process whose
-	// syscall we are processing. But since sysbox-fs did listxattr() as root, we
-	// must first filter out any xattr which the process performing the syscall
-	// has no permissions to list.
-
-	if si.size > 0 {
-
-		// Filter out trusted.* if the process does not have CAP_SYS_ADMIN set.
-		if !process.IsCapabilitySet(cap.EFFECTIVE, cap.CAP_SYS_ADMIN) {
-			resp.Val = filterXattrList(resp.Val, "trusted.")
-		}
-
-		// Filter out security.capability.* if the process does not have CAP_SETFCAP set.
-		if !process.IsCapabilitySet(cap.EFFECTIVE, cap.CAP_SETFCAP) {
-			resp.Val = filterXattrList(resp.Val, "security.capability")
-		}
-
-		if err := t.WriteRetVal(si.pid, si.addr, resp.Val); err != nil {
+	// syscall we are processing.
+	//
+	// Notice that the nsexec process executing this syscall in the container
+	// namespaces, is adopting the 'personality' of the original user process.
+	// Thus, there's no need to adjust the list of received xattrs as these ones
+	// should match those obtained if Sysbox were not in the picture.
+	//
+	// Also, bear in mind that the list of returned xattrs during the second
+	// listxattr(), may reflect a subset (or an empty set) of those reported in
+	// the first litstxattr() syscall, as it's only during the second one (i.e.,
```

Comment:
I find it strange that the kernel only checks the process caps on the second call to listxattr() but not the first one (where size = 0). The whole point of the first listxattr() call is to get the exact number of bytes that the second call would return, so one would think that the kernel would look at the process caps in both.

This aside though, the solution you have is correct: we should use resp.Size as the "size" when writing back to the process memory, because there is no guarantee that resp.Size will equal si.Size (the kernel may return less data if for example the xattr for the file changed in between the first and second listxattr() syscalls).

Reply:
> I find it strange that the kernel only checks the process caps on the second call to listxattr() but not the first one (where size = 0). The whole point of the first listxattr() call is to get the exact number of bytes that the second call would return, so one would think that the kernel would look at the process caps in both.

Yep, I was surprised to see this too, but that's exactly what's happening as you can see below (upon arrival of the second syscall below): 

```
llistxattr("/nix/store/nk6b2ckznjic5wj8ddw0wgdrn4mbz3lg-patch-registry-deps", NULL, 0) = 23
llistxattr("/nix/store/nk6b2ckznjic5wj8ddw0wgdrn4mbz3lg-patch-registry-deps", "", 23) = 0
```

```
> github.com/nestybox/sysbox-fs/seccomp.(*listxattrSyscallInfo).processListxattr() ./sysbox-fs/seccomp/xattr.go:414 (hits goroutine(121):2 total:2) (PC: 0xd51aae)
Warning: listing may not match stale executable
   409:		// must first filter out any xattr which the process performing the syscall
   410:		// has no permissions to list.
   411:
   413:		if si.size > 0 && resp.Size > 0 {
=> 414:			if err := t.WriteRetVal(si.pid, si.addr, resp.Val); err != nil {
   415:				sysResp := t.createErrorResponse(si.reqId, syscall.ENOTSUP)
   416:				return sysResp, nil
   417:			}
   418:		}
   419:

(dlv) p si
*github.com/nestybox/sysbox-fs/seccomp.listxattrSyscallInfo {
	syscallCtx: github.com/nestybox/sysbox-fs/seccomp.syscallCtx {
		syscallNum: 195,
		syscallName: "llistxattr",
		reqId: 16518291531813794632,
		pid: 27919,
		uid: 0,
		gid: 0,
		cwd: "",
		root: "",
		processInfo: github.com/nestybox/sysbox-fs/domain.ProcessIface nil,
		cntr: github.com/nestybox/sysbox-fs/domain.ContainerIface(*github.com/nestybox/sysbox-fs/state.container) ...,
		tracer: *(*"github.com/nestybox/sysbox-fs/seccomp.syscallTracer")(0xc0000bb6e0),},
	pathFd: 0,
	path: "/nix/store/11xpmmwy95396nkhih3qc3814lqhqb8f-libunistring-0.9.10",
	addr: 140104651776544,
	size: 23,}
(dlv)

(dlv) p resp
github.com/nestybox/sysbox-fs/domain.ListxattrRespPayload {
	Val: []uint8 len: 23, cap: 24, [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
	Size: 0,}
(dlv)
```

> This aside though, the solution you have is correct: we should use resp.Size as the "size" when writing back to the process memory, because there is no guarantee that resp.Size will equal si.Size (the kernel may return less data if for example the xattr for the file changed in between the first and second listxattr() syscalls).

Agree, that's another reason to rely on resp.Size other than the one I mentioned above.